### PR TITLE
Tests - Adding route tests for user/admin CRUD operations

### PR DESCRIPTION
**Pull request by lirantal**

### CODE REVIEW

File: modules/users/tests/server/user.server.routes.tests.js [hunk 1]
```diff
@@ -40,7 +40,9 @@ describe('User CRUD tests', function () {
       username: credentials.username,
       password: credentials.password,
       provider: 'local'
-    });
+    };
+
+    user = new User(_user);
 
     // Save a user to the test db and create new article
     user.save(function () {
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: Shouldn't we check that the user was saved and an error wasn't thrown?

File: modules/users/tests/server/user.server.routes.tests.js [hunk 2]
```diff
@@ -48,6 +50,56 @@ describe('User CRUD tests', function () {
...
+        if (signupErr) {
+          return done(signupErr);
+        }
+
+        signupRes.body.username.should.equal(_user.username);
+        signupRes.body.email.should.equal(_user.email);
+        // Assert a proper profile image has been set, even if by default
+        signupRes.body.profileImageURL.should.not.be.empty();
+        // Assert we have just the default 'user' role
+        signupRes.body.roles.should.be.instanceof(Array).and.have.lengthOf(1);
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: This doesn't check the role, it just checks length.

File: modules/users/tests/server/user.server.routes.tests.js [hunk 3]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
     });
   });
 
+  it('should be able to get a single user details if admin', function (done) {
+    user.roles = ['user', 'admin'];
+
+    user.save(function () {
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: Shouldn't we check to make sure it was saved properly?

File: modules/users/tests/server/user.server.routes.tests.js [hunk 4]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
...
+              done();
+            });
+        });
+    });
+  });
+
+  it('should be able to update a single user details if admin', function (done) {
+    user.roles = ['user', 'admin'];
+
+    user.save(function () {
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: check?

File: modules/users/tests/server/user.server.routes.tests.js [hunk 5]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
...
+            .end(function (userInfoErr, userInfoRes) {
+              if (userInfoErr) {
+                return done(userInfoErr);
+              }
+
+              userInfoRes.body.should.be.instanceof(Object);
+              userInfoRes.body._id.should.be.equal(String(user._id));
+
+              // Call the assertion callback
+              done();
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: return?

File: modules/users/tests/server/user.server.routes.tests.js [hunk 6]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
...
+              }
+
+              userInfoRes.body.should.be.instanceof(Object);
+              userInfoRes.body.firstName.should.be.equal('admin_update_first');
+              userInfoRes.body.lastName.should.be.equal('admin_update_last');
+              userInfoRes.body.roles.should.be.instanceof(Array).and.have.lengthOf(1);
+              userInfoRes.body._id.should.be.equal(String(user._id));
+
+              // Call the assertion callback
+              done();
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: return?

File: modules/users/tests/server/user.server.routes.tests.js [hunk 7]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
...
+            });
+        });
+    });
+  });
+
+  it('should be able to delete a single user if admin', function (done) {
+    user.roles = ['user', 'admin'];
+
+    user.save(function () {
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: check

File: modules/users/tests/server/user.server.routes.tests.js [hunk 8]
```diff
@@ -101,6 +153,110 @@ describe('User CRUD tests', function () {
...
+            .end(function (userInfoErr, userInfoRes) {
+              if (userInfoErr) {
+                return done(userInfoErr);
+              }
+
+              userInfoRes.body.should.be.instanceof(Object);
+              userInfoRes.body._id.should.be.equal(String(user._id));
+
+              // Call the assertion callback
+              done();
```
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: return?

File: modules/users/tests/server/user.server.routes.tests.js (hunk 4 above)
Comment (lirantal) [inline modules/users/tests/server/user.server.routes.tests.js]: you mean to check user.save() is ok?

File: modules/users/tests/server/user.server.routes.tests.js (hunk 5 above)
Comment (lirantal) [inline modules/users/tests/server/user.server.routes.tests.js]: the test anyway ends here so it doesn't matter that much, right?
I anyway updated to return here so it is consistent with the rest of the tests

**Comment by lirantal** [comment]
@ilanbiala thanks, great feedback!
I addressed all the comments.

File: modules/users/tests/server/user.server.routes.tests.js (hunk 4 above)
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: Yes

**Comment by lirantal** [comment]
@ilanbiala any ideas why I don't see the coverage status here? did something break there?
We do see it for other PRs for example:
image

**Comment by ilanbiala** [comment]
@lirantal I noticed that, I'm not sure exactly why but Coveralls had the last build as 1352. I just restarted the build so hopefully Coveralls will get that one.

**Comment by lirantal** [comment]
Yep, now it's there, thanks.

**Comment by lirantal** [comment]
@ilanbiala updated and addressed all your comments.

File: modules/users/tests/server/user.server.routes.tests.js (hunk 2 above)
Comment (ilanbiala) [inline modules/users/tests/server/user.server.routes.tests.js]: Resolved with L74.

**Comment by ilanbiala** [comment]
@lirantal one more return missed https://github.com/meanjs/mean/pull/928/files#diff-f2da820fbb902b375a9e67035ac86c1bR152

**Comment by lirantal** [comment]
missed it, thanks!

**Comment by ilanbiala** [comment]
LGTM.

**Comment by codydaig** [comment]
LGTM

**Pull request merged**